File: app/test/unit/features/admissions/domain/applicant_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:logicclass/core/constants/education_level.dart';
import 'package:logicclass/features/admissions/domain/entities/applicant.dart';

/// The pipeline a private school's year is won or lost in.
///
/// What is being tested here is mostly not arithmetic. It is that the
/// stages cannot be set to whatever somebody expects, that a family who
/// went elsewhere is counted apart from one the school turned down, and
/// that the list of people to ring this morning is actually the people
/// nobody has rung.
void main() {
  Applicant applicant({
    String id = 'app_1',
    AdmissionStage stage = AdmissionStage.inquiry,
    DateTime? stageChangedAt,
    String? studentId,
    double? examScore,
    double? examMaxScore,
  }) =>
      Applicant(
        id: id,
        referenceNumber: 'A-0001',
        firstName: 'Bea',
        lastName: 'Torres',
        educationLevel: EducationLevel.highSchool,
        gradeLevel: 'Grade 7',
        guardianName: 'Rosario Torres',
        guardianPhone: '09171234567',
        stage: stage,
        inquiredAt: DateTime(2026, 1, 10),
        stageChangedAt: stageChangedAt ?? DateTime(2026, 1, 10),
        examScore: examScore,
        examMaxScore: examMaxScore,
        studentId: studentId,
      );

  group('where an applicant may be moved to', () {
    test('forward one step, back one step, or out', () {
      final next = nextStagesFrom(AdmissionStage.examScheduled);
      expect(next, contains(AdmissionStage.examTaken));
      expect(next, contains(AdmissionStage.applied));
      expect(next, contains(AdmissionStage.declined));
      expect(next, contains(AdmissionStage.withdrawn));
    });

    test('never several steps forward at once', () {
      // A pipeline whose stages can be set freely stops meaning
      // anything: somebody marks a family as offered because that is the
      // outcome they expect, and the funnel then reports offers the
      // school never made.
      expect(nextStagesFrom(AdmissionStage.inquiry), isNot(contains(AdmissionStage.offered)));
      expect(nextStagesFrom(AdmissionStage.inquiry), isNot(contains(AdmissionStage.enrolled)));
    });

    test('the first stage has nowhere to go back to', () {
      expect(nextStagesFrom(AdmissionStage.inquiry), [
        AdmissionStage.applied,
        AdmissionStage.declined,
        AdmissionStage.withdrawn,
      ]);
    });

    test('a family who came back starts again rather than resuming', () {
      expect(nextStagesFrom(AdmissionStage.withdrawn), [AdmissionStage.inquiry]);
      expect(nextStagesFrom(AdmissionStage.declined), [AdmissionStage.inquiry]);
    });

    test('nothing moves out of enrolled', () {
      // There is a student record behind it. Moving out would leave a
      // child enrolled in the school and an applicant record saying they
      // withdrew.
      expect(nextStagesFrom(AdmissionStage.enrolled), isEmpty);
    });
  });

  group('the entrance exam', () {
    test('is a percentage of what it was out of', () {
      final sat = applicant(examScore: 68, examMaxScore: 80);
      expect(sat.examPercentage, 85.0);
    });

    test('is nothing at all until it is sat', () {
      // Not zero. A child who has not taken the test has not failed it,
      // and a list sorted by score would put them below everybody who
      // did badly.
      expect(applicant().examPercentage, isNull);
      expect(applicant(examScore: 40).examPercentage, isNull);
    });

    test('an exam out of nothing is not a score', () {
      expect(applicant(examScore: 40, examMaxScore: 0).examPercentage, isNull);
    });
  });

  group('who to ring this morning', () {
    final today = DateTime(2026, 2, 20);

    test('open applicants nobody has moved in a week, longest wait first', () {
      final list = applicantsNeedingFollowUp(
        [
          applicant(id: 'a', stageChangedAt: DateTime(2026, 2, 1)),
          applicant(id: 'b', stageChangedAt: DateTime(2026, 2, 12)),
          applicant(id: 'c', stageChangedAt: DateTime(2026, 2, 19)),
        ],
        asOf: today,
      );
      expect(list.map((a) => a.id), ['a', 'b']);
    });

    test('leaves out the ones who are finished, however long ago', () {
      // An enrolled family is not waiting for a phone call, and a list
      // that says they are is a list the office stops reading.
      final list = applicantsNeedingFollowUp(
        [
          applicant(
              id: 'enrolled',
              stage: AdmissionStage.enrolled,
              studentId: 'stu_1',
              stageChangedAt: DateTime(2025, 8, 1)),
          applicant(
              id: 'gone',
              stage: AdmissionStage.withdrawn,
              stageChangedAt: DateTime(2025, 8, 1)),
          applicant(id: 'waiting', stageChangedAt: DateTime(2026, 2, 1)),
        ],
        asOf: today,
      );
      expect(list.map((a) => a.id), ['waiting']);
    });

    test('counts days in the stage, not days since the enquiry', () {
      // A family being actively worked through the stages is not going
      // cold, however long ago they first rang.
      final busy = applicant(
        stage: AdmissionStage.offered,
        stageChangedAt: DateTime(2026, 2, 19),
      );
      expect(busy.daysInStage(today), 1);
      expect(applicantsNeedingFollowUp([busy], asOf: today), isEmpty);
    });
  });

  group('the funnel', () {
    final applicants = [
      applicant(id: '1'),
      applicant(id: '2'),
      applicant(id: '3', stage: AdmissionStage.applied),
      applicant(id: '4', stage: AdmissionStage.offered),
      applicant(id: '5', stage: AdmissionStage.enrolled, studentId: 'stu_5'),
      applicant(id: '6', stage: AdmissionStage.enrolled, studentId: 'stu_6'),
      applicant(id: '7', stage: AdmissionStage.declined),
      applicant(id: '8', stage: AdmissionStage.withdrawn),
    ];

    test('counts who is where right now', () {
      final funnel = AdmissionFunnel.of(applicants);
      expect(funnel.nowIn[AdmissionStage.inquiry], 2);
      expect(funnel.nowIn[AdmissionStage.enrolled], 2);
      expect(funnel.total, 8);
    });

    test('counts who has got at least this far', () {
      // Both numbers are needed. Forty families are at the enquiry stage
      // today; four hundred passed through it this year, and a report
      // giving only the first looks like a school with no pipeline at
      // all by June.
      final funnel = AdmissionFunnel.of(applicants);
      // Everybody still on the pipeline: the two at enquiry, the one who
      // applied, the one offered, and the two enrolled. The declined and
      // the withdrawn are not on it -- see the comment on the factory.
      expect(funnel.reached[AdmissionStage.inquiry], 6);
      expect(funnel.reached[AdmissionStage.offered], 3);
      expect(funnel.reached[AdmissionStage.enrolled], 2);
    });

    test('keeps "we said no" apart from "they went elsewhere"', () {
      // Different problems, and only one of them is the school's doing.
      final funnel = AdmissionFunnel.of(applicants);
      expect(funnel.declined, 1);
      expect(funnel.withdrawn, 1);
      expect(funnel.open, 4);
    });

    test('converts enrolments against everybody who ever enquired', () {
      expect(AdmissionFunnel.of(applicants).conversionRate, 25.0);
    });

    test('a stage nobody has reached has no rate, rather than zero', () {
      // 0% for a stage nobody has got to yet is a report that reads as a
      // disaster when nothing has happened at all.
      final early = AdmissionFunnel.of([applicant(id: '1')]);
      expect(early.rateBetween(AdmissionStage.offered, AdmissionStage.enrolled), isNull);
      expect(early.conversionRate, 0.0);
    });

    test('an empty year has no conversion rate at all', () {
      expect(AdmissionFunnel.of(const []).conversionRate, isNull);
    });
  });

  test('an applicant with a student behind them has enrolled', () {
    expect(applicant(stage: AdmissionStage.enrolled, studentId: 'stu_1').hasEnrolled, isTrue);
    expect(applicant().hasEnrolled, isFalse);
  });
}

File: functions/test/shared/admissions/applicant.test.ts
import {
  nextStagesFrom,
  requireLegalTransition,
  validateApplicant,
  validateExamResult,
  validateReservationFee,
} from "../../../src/shared/admissions/applicant";

/**
 * The server's half of the admissions pipeline.
 *
 * The client offers the right buttons; this is what holds when something
 * else calls the callable. The rule worth defending is that stages
 * cannot be set to whatever somebody expects -- a pipeline whose stages
 * are free-for-all stops meaning anything within a term, and the funnel
 * starts reporting offers the school never made.
 */
describe("nextStagesFrom", () => {
  it("offers one step forward, one back, and both endings", () => {
    expect(nextStagesFrom("exam_scheduled")).toEqual([
      "exam_taken",
      "applied",
      "declined",
      "withdrawn",
    ]);
  });

  it("has nowhere to go back to from the first stage", () => {
    expect(nextStagesFrom("inquiry")).toEqual(["applied", "declined", "withdrawn"]);
  });

  it("sends a family who came back to the beginning", () => {
    expect(nextStagesFrom("withdrawn")).toEqual(["inquiry"]);
    expect(nextStagesFrom("declined")).toEqual(["inquiry"]);
  });

  it("lets nothing out of enrolled", () => {
    // There is a student record behind it. Moving out would leave a
    // child enrolled and an applicant record saying they withdrew.
    expect(nextStagesFrom("enrolled")).toEqual([]);
  });
});

describe("requireLegalTransition", () => {
  it("allows the next step", () => {
    expect(() => requireLegalTransition("applied", "exam_scheduled")).not.toThrow();
  });

  it("allows a step back, so a mis-marked family can be corrected", () => {
    // Without this, people fix a wrong stage by making a second record
    // for the same child.
    expect(() => requireLegalTransition("offered", "exam_taken")).not.toThrow();
  });

  it("refuses a jump, and names what is allowed instead", () => {
    expect(() => requireLegalTransition("inquiry", "offered"))
      .toThrow(/can only move to: applied, declined, withdrawn/);
  });

  it("refuses moving somewhere they already are", () => {
    expect(() => requireLegalTransition("applied", "applied"))
      .toThrow(/already where they are/);
  });

  it("refuses setting enrolled at all", () => {
    // Enrolment creates the student record, so it is reached by
    // enrolling and not by setting a stage. An applicant marked enrolled
    // with no record is a child the school believes is enrolled and the
    // registrar cannot find.
    expect(() => requireLegalTransition("reserved", "enrolled"))
      .toThrow(/done by enrolling them rather than by setting the stage/);
  });

  it("refuses moving an enrolled applicant anywhere", () => {
    expect(() => requireLegalTransition("enrolled", "withdrawn"))
      .toThrow(/cannot be moved back out of it/);
  });
});

describe("validateApplicant", () => {
  const enquiry = {
    firstName: "Bea",
    lastName: "Marquez",
    educationLevel: "high_school",
    gradeLevel: "Grade 7",
    guardianName: "Alma Marquez",
    guardianPhone: "09171234567",
  };

  it("takes down an enquiry with the little it insists on", () => {
    const saved = validateApplicant(enquiry);
    expect(saved.firstName).toBe("Bea");
    expect(saved.gradeLevel).toBe("Grade 7");
    // Absent optional fields are stored as null, not "", so a later read
    // can tell "not asked" from "asked and blank".
    expect(saved.middleName).toBeNull();
    expect(saved.source).toBeNull();
  });

  it("insists on a name", () => {
    expect(() => validateApplicant({...enquiry, lastName: " "}))
      .toThrow(/name is required/);
  });

  it("insists on somebody to ring", () => {
    // An applicant nobody can contact is not a lead, it is a row.
    expect(() => validateApplicant({...enquiry, guardianPhone: ""}))
      .toThrow(/not a lead/);
    expect(() => validateApplicant({...enquiry, guardianName: ""}))
      .toThrow(/not a lead/);
  });

  it("insists the number is one that could actually be rung", () => {
    // The field was mandatory but unchecked, so "0" satisfied it. The
    // whole justification for making it mandatory is that somebody can
    // be rung back, and a row with "0" in it is the row this validator
    // exists to refuse, wearing a disguise.
    for (const bad of ["0", "12", "asdf", "555-0100", "n/a"]) {
      expect(() => validateApplicant({...enquiry, guardianPhone: bad}))
        .toThrow(/not a mobile number this system can read/);
    }
  });

  it("accepts every shape a Philippine number is written in", () => {
    for (const shape of [
      "09171234567",
      "+639171234567",
      "9171234567",
      "0917 123 4567",
      "(0917) 123-4567",
    ]) {
      const saved = validateApplicant({...enquiry, guardianPhone: shape});
      // Stored as typed. The office reads this back to a family.
      expect(saved.guardianPhone).toBe(shape);
    }
  });

  it("refuses a guardian address that would be copied onto a student", () => {
    // At enrolment this lands on the student record as a guardian
    // contact. An address accepted here is one the student form and the
    // student import would both have refused -- admissions was the back
    // door around them.
    expect(() => validateApplicant({...enquiry, guardianEmail: "alma@gmailcom"}))
      .toThrow(/not a valid email address for the guardian/);
    expect(validateApplicant({...enquiry, guardianEmail: "  Alma@Gmail.com "}).guardianEmail)
      .toBe("alma@gmail.com");
  });

  it("takes the applicant's own email and number, and refuses a bad one", () => {
    // A Grade 1 applicant has neither, which is why they are optional. A
    // Senior High applicant has both, and they are the person the school
    // is actually talking to.
    const none = validateApplicant(enquiry);
    expect(none.email).toBeNull();
    expect(none.phone).toBeNull();

    const both = validateApplicant({
      ...enquiry,
      email: "Bea.Marquez@student.school.edu.ph",
      phone: "+63 918 555 0100",
    });
    expect(both.email).toBe("bea.marquez@student.school.edu.ph");
    expect(both.phone).toBe("+63 918 555 0100");

    expect(() => validateApplicant({...enquiry, email: "bea@gmailcom"}))
      .toThrow(/not a valid email address for the applicant/);
    expect(() => validateApplicant({...enquiry, phone: "12"}))
      .toThrow(/not a mobile number this system can read/);
  });

  it("insists on which year they are applying into", () => {
    expect(() => validateApplicant({...enquiry, gradeLevel: ""}))
      .toThrow(/year they are applying into/);
  });

  it("asks for nothing else, because the caller is still on the line", () => {
    // A form that refuses to save without a middle name is a form that
    // gets abandoned -- and then the enquiry is a note on paper again.
    expect(() => validateApplicant(enquiry)).not.toThrow();
  });
});

describe("validateExamResult", () => {
  it("takes a score and what it was out of", () => {
    expect(validateExamResult(68, 80)).toEqual({score: 68, maxScore: 80});
  });

  it("refuses a score above the paper", () => {
    // Almost always the two fields the wrong way round, and it would
    // rank a child above everybody who sat the same paper.
    expect(() => validateExamResult(80, 68)).toThrow(/right way round/);
  });

  it("refuses a paper out of nothing", () => {
    expect(() => validateExamResult(40, 0)).toThrow(/more than zero/);
  });

  it("refuses a negative or unreadable score", () => {
    expect(() => validateExamResult(-1, 80)).toThrow(/not a number/);
    expect(() => validateExamResult("absent", 80)).toThrow(/not a number/);
  });

  it("allows a zero, which is a real mark", () => {
    expect(validateExamResult(0, 80).score).toBe(0);
  });
});

describe("validateReservationFee", () => {
  it("rounds to centavos", () => {
    expect(validateReservationFee(2000.005)).toBe(2000.01);
  });

  it("refuses nothing, or less than nothing", () => {
    expect(() => validateReservationFee(0)).toThrow(/above zero/);
    expect(() => validateReservationFee(-500)).toThrow(/above zero/);
    expect(() => validateReservationFee("free")).toThrow(/above zero/);
  });
});
